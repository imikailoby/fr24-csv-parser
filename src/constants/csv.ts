export enum CsvDataRowIndex {
  HEADER = 0,
  FIRST_BODY_ROW,
}

export enum CsvDataColumnIndex {
  DATE = 0,
  FLIGHT_NUMBER,
  FROM,
  TO,
  DEP_TIME,
  ARR_TIME,
  DURATION,
  AIRLINE,
  AIRCRAFT,
  REGISTRATION,
  SEAT_NUMBER,
  SEAT_TYPE,
  FLIGHT_CLASS,
  FLIGHT_REASON,
  NOTE,
  DEP_ID,
  ARR_ID,
  AIRLINE_ID,
  AIRCRAFT_ID,
}

export const EXPECTED_CSV_COLUMNS = [
  'Date',
  'Flight number',
  'From',
  'To',
  'Dep time',
  'Arr time',
  'Duration',
  'Airline',
  'Aircraft',
  'Registration',
  'Seat number',
  'Seat type',
  'Flight class',
  'Flight reason',
  'Note',
  'Dep_id',
  'Arr_id',
  'Airline_id',
  'Aircraft_id',
];
