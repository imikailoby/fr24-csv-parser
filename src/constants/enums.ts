export enum SeatType {
  WINDOW = 'window',
  MIDDLE = 'middle',
  AISLE = 'aisle',
}

export enum CabinClass {
  PRIVATE = 'private',
  FIRST = 'first',
  BUSINESS = 'business',
  ECONOMY_PLUS = 'economyPlus',
  ECONOMY = 'economy',
}

export enum FlightReason {
  OTHER = 'other',
  CREW = 'crew',
  BUSINESS = 'business',
  LEISURE = 'leisure',
}
