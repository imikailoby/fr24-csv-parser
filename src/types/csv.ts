export interface ParsedCsvRow {
  date: string;
  flightNumber: string;
  from: string;
  to: string;
  departureTime: string;
  arrivalTime: string;
  duration: string;
  airline: string;
  aircraft: string;
  registration: string;
  seatNumber: string;
  seatType: string;
  flightClass: string;
  flightReason: string;
  note: string;
  departureId: string;
  arrivalId: string;
  airlineId: string;
  aircraftId: string;
}
