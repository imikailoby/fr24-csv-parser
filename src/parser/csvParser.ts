import { CsvDataRowIndex } from '../constants/csv';
import { cleanValue, detectDelimiter } from '../utils/csvUtils';

// First line is always header, second and following are data
const MIN_LINES_WITH_HEADER_AND_DATA = 2;

export function parseCsv(csvText: string): string[][] {
  const delimiter = detectDelimiter(csvText);
  const lines = csvText.split(/\r?\n/).filter((line) => line.trim() !== '');

  if (lines.length < MIN_LINES_WITH_HEADER_AND_DATA) {
    return [];
  }

  return lines.map((line) => line.split(delimiter).map((value) => cleanValue(value.trim())));
}

export function validateCsvColumns(csvText: string, expectedColumns: string[]): void {
  const lines = csvText.split(/\r?\n/).filter((line) => line.trim() !== '');
  if (lines.length === 0) {
    throw new Error('CSV-file is empty');
  }

  const headers = lines[CsvDataRowIndex.HEADER].split(',').map((h) => h.trim().replace(/^"(.*)"$/, '$1'));

  if (headers.length !== expectedColumns.length) {
    throw new Error(`Expected ${expectedColumns.length} columns, but got ${headers.length}`);
  }

  for (let i = 0; i < expectedColumns.length; i++) {
    if (headers[i] !== expectedColumns[i]) {
      throw new Error(`Error in the ${i + 1} column: expected "${expectedColumns[i]}", but got "${headers[i]}"`);
    }
  }
}
