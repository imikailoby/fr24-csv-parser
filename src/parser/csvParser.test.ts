import { EXPECTED_CSV_COLUMNS } from '../constants/csv';
import {
  mockedCsvContent,
  mockedCsvContentInvalid,
  mockedCsvContentInvalidHeaders,
  mockedCsvContentEmpty,
} from '../tests/mocks';
import { parseCsv, validateCsvColumns } from './csvParser';

describe('csvParser', () => {
  it('parseCsv', () => {
    expect(parseCsv(mockedCsvContent)).toMatchSnapshot('valid-snapshot');
    expect(parseCsv(mockedCsvContentEmpty)).toMatchSnapshot('empty-snapshot');
  });

  it('validateCsvColumns', () => {
    function emptyFunction() {
      validateCsvColumns('', EXPECTED_CSV_COLUMNS);
    }
    function invalidFunction() {
      validateCsvColumns(mockedCsvContentInvalid, EXPECTED_CSV_COLUMNS);
    }
    function invalidHeadersFunction() {
      validateCsvColumns(mockedCsvContentInvalidHeaders, EXPECTED_CSV_COLUMNS);
    }

    expect(validateCsvColumns(mockedCsvContent, EXPECTED_CSV_COLUMNS)).toBeUndefined();
    expect(emptyFunction).toThrowErrorMatchingSnapshot('empty-snapshot');
    expect(invalidFunction).toThrowErrorMatchingSnapshot('invalid-snapshot');
    expect(invalidHeadersFunction).toThrowErrorMatchingSnapshot('invalid-headers-snapshot');
  });
});
