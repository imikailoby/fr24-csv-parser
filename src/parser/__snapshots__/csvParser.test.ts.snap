// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csvParser parseCsv: empty-snapshot 1`] = `[]`;

exports[`csvParser parseCsv: valid-snapshot 1`] = `
[
  [
    "Date",
    "Flight number",
    "From",
    "To",
    "Dep time",
    "Arr time",
    "Duration",
    "Airline",
    "Aircraft",
    "Registration",
    "Seat number",
    "Seat type",
    "Flight class",
    "Flight reason",
    "Note",
    "Dep_id",
    "Arr_id",
    "Airline_id",
    "Aircraft_id",
  ],
  [
    "2024-10-12",
    "DL49",
    "Amsterdam / Schiphol (AMS/EHAM)",
    "New York / John F Kennedy (JFK/KJFK)",
    "12:10:00",
    "14:30:00",
    "08:20:00",
    "Delta Air Lines (DL/DAL)",
    "Airbus A330-900neo (A339)",
    "",
    "1A",
    "1",
    "1",
    "1",
    "",
    "122",
    "1328",
    "223",
    "2130",
  ],
]
`;

exports[`csvParser validateCsvColumns: empty-snapshot 1`] = `"CSV-file is empty"`;

exports[`csvParser validateCsvColumns: invalid-headers-snapshot 1`] = `"Error in the 2 column: expected "Flight number", but got """`;

exports[`csvParser validateCsvColumns: invalid-snapshot 1`] = `"Expected 19 columns, but got 18"`;
